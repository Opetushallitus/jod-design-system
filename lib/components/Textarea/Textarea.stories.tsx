import type { Meta, StoryObj } from '@storybook/react';
import { fn } from '@storybook/test';
import React from 'react';

import { Textarea } from './Textarea';

const meta = {
  title: 'Forms/Textarea',
  component: Textarea,
  tags: ['autodocs'],
} satisfies Meta<typeof Textarea>;

export default meta;

type Story = StoryObj<typeof meta>;

export const Default: Story = {
  render: (args: Story['args']) => {
    const { value, onChange, ...rest } = args;
    const [textareaValue, setTextareaValue] = React.useState(value);
    return (
      <Textarea
        value={textareaValue}
        onChange={(event) => {
          setTextareaValue(event.target.value);
          onChange(event);
        }}
        {...rest}
      />
    );
  },
  decorators: [
    (Story) => (
      <div className="ds-max-w-[480px]">
        <Story />
      </div>
    ),
  ],
  parameters: {
    design: {
      type: 'figma',
      url: 'https://www.figma.com/design/6M2LrpSCcB0thlFDaQAI2J/cx_jod_client?node-id=5805-3164',
    },
    docs: {
      description: {
        story: 'This is a input field component for displaying and editing a value with a label text.',
      },
    },
  },
  args: {
    value: 'Lorem ipsum dolor sit amet',
    onChange: fn(),
    label: 'Label text',
  },
};
