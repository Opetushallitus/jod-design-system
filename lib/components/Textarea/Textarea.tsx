import React from 'react';
import { tidyClasses as tc } from '../../utils';

interface BaseTextareaProps {
  /** The name of the textarea */
  name?: string;
  /** The value of the textarea */
  value?: string;
  /** The function to call when the textarea loses focus */
  onBlur?: (event: React.FocusEvent<HTMLTextAreaElement>) => void;
  /** The function to call when the value of the textarea changes */
  onChange: (event: React.ChangeEvent<HTMLTextAreaElement>) => void;
  /** The function to call when a key is pressed in the textarea */
  onKeyDown?: (event: React.KeyboardEvent<HTMLTextAreaElement>) => void;
  /** The placeholder text to display in the textarea */
  placeholder?: string;
  /** The help text to display below the textarea */
  help?: string;
  /** The maximum number of characters that can be entered into the textarea */
  maxLength?: number;
  /** The number of rows to display in the textarea */
  rows?: number;
  /** Additional classes to add to the textarea */
  className?: string;
}

interface HideLabelProps extends BaseTextareaProps {
  /** The label text is not shown when hideLabel is true */
  label?: never;
  /** Hide label */
  hideLabel?: true;
  /** The placeholder text to display in the textarea is required when hideLabel is true */
  placeholder: string;
}
interface ShowLabelProps extends BaseTextareaProps {
  /** The label text to display above the textarea */
  label: string;
  /** Hide label */
  hideLabel?: false;
  /** The placeholder text to display in the textarea */
  placeholder?: string;
}

export type TextareaProps = ShowLabelProps | HideLabelProps;

/** Textareas are multi-line text boxes that allow users to input custom text entries with a keyboard. Various options can be shown with the field to communicate the input requirements. */
export const Textarea = React.forwardRef<HTMLTextAreaElement, TextareaProps>(function Textarea(
  {
    name,
    value,
    onBlur,
    onChange,
    onKeyDown,
    placeholder,
    label,
    hideLabel = false,
    help,
    maxLength,
    rows,
    className = '',
  }: TextareaProps,
  ref,
) {
  const inputId = React.useId();
  const helpId = React.useId();
  return (
    <>
      <label
        htmlFor={inputId}
        className={tc([
          hideLabel ? 'ds-hidden' : '',
          'ds-mb-4 ds-inline-block ds-align-top ds-text-form-label ds-font-arial ds-text-black',
        ])}
      >
        {label}
      </label>
      <textarea
        ref={ref}
        id={inputId}
        name={name}
        value={value}
        onBlur={onBlur}
        onChange={onChange}
        onKeyDown={onKeyDown}
        maxLength={maxLength}
        rows={rows}
        placeholder={placeholder}
        autoComplete="off"
        aria-describedby={help ? helpId : undefined}
        className={tc([
          'ds-block ds-w-full ds-rounded ds-border ds-border-border-gray ds-bg-white ds-p-5 ds-text-black ds-outline-none placeholder:ds-text-secondary-gray ds-font-arial ds-text-body-md',
          className,
        ])}
      />
      {help && (
        <div id={helpId} className="ds-mt-2 ds-block ds-text-help ds-text-secondary-gray ds-font-arial">
          {help}
        </div>
      )}
    </>
  );
});
