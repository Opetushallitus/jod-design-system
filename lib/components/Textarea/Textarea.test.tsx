import { fireEvent, render, screen } from '@testing-library/react';
import { describe, expect, it, vi } from 'vitest';
import { Textarea } from './Textarea';

describe('Textarea', () => {
  const placeholder = 'Enter text';

  it('renders without crashing', () => {
    const onChange = vi.fn();
    render(<Textarea onChange={onChange} placeholder={placeholder} dataTestId="ta1" />);
    expect(screen.getByTestId('ta1')).toBeInTheDocument();
    const textarea = screen.getByPlaceholderText(placeholder);
    expect(textarea).toBeInTheDocument();
    expect(screen.getByTestId('ta1')).toBeInTheDocument();
  });

  it('displays the label when hideLabel is false', () => {
    const onChange = vi.fn();
    render(<Textarea onChange={onChange} label="Test Label" placeholder="Enter text" dataTestId="ta2" />);
    expect(screen.getByTestId('ta2')).toBeInTheDocument();
    const label = screen.getByText('Test Label');
    expect(label).toBeInTheDocument();
    expect(screen.getByTestId('ta2')).toBeInTheDocument();
  });

  it('hides the label when hideLabel is true', () => {
    const onChange = vi.fn();
    render(<Textarea onChange={onChange} hideLabel={true} placeholder="Enter text" dataTestId="ta3" />);
    expect(screen.getByTestId('ta3')).toBeInTheDocument();
    const label = screen.queryByText('Test Label');
    expect(label).not.toBeInTheDocument();
    expect(screen.getByTestId('ta3')).toBeInTheDocument();
  });

  it('calls onChange when the value changes', () => {
    const onChange = vi.fn();
    render(<Textarea onChange={onChange} placeholder="Enter text" dataTestId="ta4" />);
    expect(screen.getByTestId('ta4')).toBeInTheDocument();
    const textarea = screen.getByPlaceholderText(placeholder);
    fireEvent.change(textarea, { target: { value: 'New value' } });
    expect(onChange).toHaveBeenCalledTimes(1);
    expect(screen.getByTestId('ta4')).toBeInTheDocument();
  });

  it('calls onBlur when the textarea loses focus', () => {
    const onChange = vi.fn();
    const onBlur = vi.fn();
    render(<Textarea onChange={onChange} onBlur={onBlur} placeholder="Enter text" dataTestId="ta5" />);
    expect(screen.getByTestId('ta5')).toBeInTheDocument();
    const textarea = screen.getByPlaceholderText(placeholder);
    fireEvent.blur(textarea);
    expect(onBlur).toHaveBeenCalledTimes(1);
    expect(screen.getByTestId('ta5')).toBeInTheDocument();
  });

  it('displays help text when provided', () => {
    const onChange = vi.fn();
    render(<Textarea onChange={onChange} help="Help text" placeholder="Enter text" dataTestId="ta6" />);
    expect(screen.getByTestId('ta6')).toBeInTheDocument();
    const helpText = screen.getByText('Help text');
    expect(helpText).toBeInTheDocument();
    expect(screen.getByTestId('ta6')).toBeInTheDocument();
  });

  it('sets the correct number of rows', () => {
    const onChange = vi.fn();
    render(<Textarea onChange={onChange} rows={5} placeholder="Enter text" dataTestId="ta7" />);
    expect(screen.getByTestId('ta7')).toBeInTheDocument();
    const textarea = screen.getByPlaceholderText(placeholder);
    expect(textarea).toHaveAttribute('rows', '5');
    expect(screen.getByTestId('ta7')).toBeInTheDocument();
  });
});
