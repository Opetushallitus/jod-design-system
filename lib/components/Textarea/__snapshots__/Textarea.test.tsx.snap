// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Textarea > calls onBlur when the textarea loses focus 1`] = `
<div>
  <label
    class="ds-mb-4 ds-inline-block ds-align-top ds-text-form-label ds-font-arial ds-text-black"
    for=":r8:"
  />
  <textarea
    autocomplete="off"
    class="ds-block ds-w-full ds-rounded ds-border ds-border-border-gray ds-bg-white ds-p-5 ds-text-black ds-outline-none placeholder:ds-text-secondary-gray ds-font-arial ds-text-body-md"
    id=":r8:"
    placeholder="Enter text"
  />
</div>
`;

exports[`Textarea > calls onChange when the value changes 1`] = `
<div>
  <label
    class="ds-mb-4 ds-inline-block ds-align-top ds-text-form-label ds-font-arial ds-text-black"
    for=":r6:"
  />
  <textarea
    autocomplete="off"
    class="ds-block ds-w-full ds-rounded ds-border ds-border-border-gray ds-bg-white ds-p-5 ds-text-black ds-outline-none placeholder:ds-text-secondary-gray ds-font-arial ds-text-body-md"
    id=":r6:"
    placeholder="Enter text"
  />
</div>
`;

exports[`Textarea > displays help text when provided 1`] = `
<div>
  <label
    class="ds-mb-4 ds-inline-block ds-align-top ds-text-form-label ds-font-arial ds-text-black"
    for=":ra:"
  />
  <textarea
    aria-describedby=":rb:"
    autocomplete="off"
    class="ds-block ds-w-full ds-rounded ds-border ds-border-border-gray ds-bg-white ds-p-5 ds-text-black ds-outline-none placeholder:ds-text-secondary-gray ds-font-arial ds-text-body-md"
    id=":ra:"
    placeholder="Enter text"
  />
  <div
    class="ds-mt-2 ds-block ds-text-help ds-text-secondary-gray ds-font-arial"
    id=":rb:"
  >
    Help text
  </div>
</div>
`;

exports[`Textarea > displays the label when hideLabel is false 1`] = `
<div>
  <label
    class="ds-mb-4 ds-inline-block ds-align-top ds-text-form-label ds-font-arial ds-text-black"
    for=":r2:"
  >
    Test Label
  </label>
  <textarea
    autocomplete="off"
    class="ds-block ds-w-full ds-rounded ds-border ds-border-border-gray ds-bg-white ds-p-5 ds-text-black ds-outline-none placeholder:ds-text-secondary-gray ds-font-arial ds-text-body-md"
    id=":r2:"
    placeholder="Enter text"
  />
</div>
`;

exports[`Textarea > hides the label when hideLabel is true 1`] = `
<div>
  <label
    class="ds-hidden ds-mb-4 ds-inline-block ds-align-top ds-text-form-label ds-font-arial ds-text-black"
    for=":r4:"
  />
  <textarea
    autocomplete="off"
    class="ds-block ds-w-full ds-rounded ds-border ds-border-border-gray ds-bg-white ds-p-5 ds-text-black ds-outline-none placeholder:ds-text-secondary-gray ds-font-arial ds-text-body-md"
    id=":r4:"
    placeholder="Enter text"
  />
</div>
`;

exports[`Textarea > renders without crashing 1`] = `
<div>
  <label
    class="ds-mb-4 ds-inline-block ds-align-top ds-text-form-label ds-font-arial ds-text-black"
    for=":r0:"
  />
  <textarea
    autocomplete="off"
    class="ds-block ds-w-full ds-rounded ds-border ds-border-border-gray ds-bg-white ds-p-5 ds-text-black ds-outline-none placeholder:ds-text-secondary-gray ds-font-arial ds-text-body-md"
    id=":r0:"
    placeholder="Enter text"
  />
</div>
`;

exports[`Textarea > sets the correct number of rows 1`] = `
<div>
  <label
    class="ds-mb-4 ds-inline-block ds-align-top ds-text-form-label ds-font-arial ds-text-black"
    for=":rc:"
  />
  <textarea
    autocomplete="off"
    class="ds-block ds-w-full ds-rounded ds-border ds-border-border-gray ds-bg-white ds-p-5 ds-text-black ds-outline-none placeholder:ds-text-secondary-gray ds-font-arial ds-text-body-md"
    id=":rc:"
    placeholder="Enter text"
    rows="5"
  />
</div>
`;
